### PR TITLE
[ADR]: Ingestion data is immutable

### PR DESCRIPTION
**Pull request by ross-p-smith**
Putting together some thoughts on whether we treat each ingestion as an immutable data set for that member - and then the domain model for that data

### CODE REVIEW

File: data-reconciliation-app/docs/adr/02-data-ingest-patterns.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        "id": "member2:A0130",
+        "company_name": "amazon"
+    }
+]
+```
+
+## Decision
+
+For the purpose of this sample, we will assume that each member ingests their full data set each time. It is the process of `Reconciliation` that calculates the consensus for that particular member. A member will only receive a resultset with the same number of records that they ingested
```
Comment (macromania) [inline data-reconciliation-app/docs/adr/02-data-ingest-patterns.md:86]: The assumption of each member ingesting the data which includes all other members' data looks heavy on data storage responsibility at each bank. 

If banks will use this api to ingest data they have and reconcile it, there wouldn't be any expectation on them to store all the data at their end. Instead, the expectation could be the network holding all the data and API reconciling it when requested.

Comment (ross-p-smith) [reply data-reconciliation-app/docs/adr/02-data-ingest-patterns.md:86]: Sorry if I have explained that badly, I think you have misunderstood. I'm not saying that a member should ingest other members data - they would not know what it was. Those json stores are how we store the data internally after they have ingested *only their* data. If we stored it as a key and and array of {id,value}, then when new data is ingested we just deleted the array for that member and upload new.

**Comment by mkcomer** [comment]
All great points... I will raise with Brent about ingest requirements to inform our decision & improvement to existing ingest model

**Review by mkcomer — APPROVED** [review]
All great points... I will raise with Brent about ingest requirements (full vs delta) to inform our decision. 

From there, we can craft stories to improve our existing ingest model. Thanks for raising these points @ross-p-smith

File: data-reconciliation-app/docs/adr/02-data-ingest-patterns.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+## Example
+
+- Member A ingests 100 records
+- Member B ingests 1000 records
+- Member A reconciles and receives 100 results back
+- Member A ingests 2000 records
+- Member A reconciles and receives 2000 results back
```
Comment (mkcomer) [inline data-reconciliation-app/docs/adr/02-data-ingest-patterns.md:94]: Of the 2000 newly ingested records by member 1, these records could be a mix of records that are completely new, or updated values on the previous 100 records submitted. We cannot assume the 100 original records are a subset of the 2000 records. 

Therefore when Member A reconciles the second time, they could receive anywhere from 2000-2100 records back

**Review by julioalex-rezende — COMMENTED** [review]
I believe file `02-data-ingest-patterns` is quite out-of-date.
Is file `06-data-mutability` replacing it or both files will exist?

File: data-reconciliation-app/docs/adr/02-data-ingest-patterns.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+Proposed
+
+## Context
+
+We need to store the data from the Ingest API in a Key/Value store (Map). We need to provide a mechanism for data to be updated or deleted after an initial ingestion. 
+
+### Full or Delta?
+- Each member supplies a complete corpus of their reference data in each ingestion. This means we would treat each members data as immutable and handle record addition/deletion by clearing that members data and creating a new Map on ingestion.
+- When data is ingested it is marked whether it is add/delete.
```
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply data-reconciliation-app/docs/adr/02-data-ingest-patterns.md:13]: Hasn't this file been deleted?

**Review by julioalex-rezende — APPROVED** [review]
ship it!

**Pull request merged**